### DISCUSSION TITLE
private enum 질문

### DISCUSSION THREAD

**Discussion by MMMIIIN** [DeveloperAcademy-POSTECH/MC2-Team5-Firefighter#120] [Show and tell]
## 사건의 발단

<img width="410" alt="image" src="https://user-images.githubusercontent.com/78677571/183788153-eed878c2-6efb-4a86-80fe-43343057291f.png">

해당 View에서 `UISlider` 부분을 모드에 따라 보였다 안보였다를 처리하고 싶었습니다.

그래서 모드를 `enum`으로 만들어야겠다 생각했습니다.

```swift
class DetailEditViewController: BaseViewController {

    private enum EditMode {
        case dateEditMode
        case infoEditMode
    }
    ...
}
```

외부에서 해당 `enum` 엔 접근하지 않기때문에 `private`으로 설정했습니다.

현재 모드를 가지고 있는 변수를 하나 만들어서 모드를 판별하고 싶었습니다.

```swift
class DetailEditViewController: BaseViewController {

    private enum EditMode {
        case dateEditMode
        case infoEditMode
    }
    var editMode: EditMode = .infoEditMode // <-- error!!
    ...
}
```

위와 같이 `editMode`란 변수를 접근레벨의 default 값인 `internal` 로 설정했습니다.

> <img width="816" alt="image" src="https://user-images.githubusercontent.com/78677571/183788708-b7cef888-e166-4693-a48b-de9d508a8192.png">

그러자 아래와 같은 에러와 마주치게 되었습니다.

<img width="980" alt="image" src="https://user-images.githubusercontent.com/78677571/183788778-ffa61ee1-97e8-4733-85e1-0efcafb3f738.png">

> Property must be declared private because its type uses a private type

해당 에러를 해결하기 위해 

<img width="456" alt="image" src="https://user-images.githubusercontent.com/78677571/183788854-a5fbbf2f-7c4f-4ec0-8b83-549522ebeafd.png">

`private`로 선언하면 해결할 수 있었습니다.

하지만 이전 View에서 해당 View로 넘어올 때 모드 변수로 통해 모드를 지정해줘야 했기에 `private`를 사용할 순 없습니다.

단순히 

<img width="382" alt="image" src="https://user-images.githubusercontent.com/78677571/183788977-af72b045-3544-402b-86fb-2ff6e714fbbf.png">

둘 다 `private`를 제외하면 문제가 없지만, 해당 `enum`을 외부에서 사용하지 않기때문에 `private`처리를 하고싶은데 위와같은 이유로 되지 않습니다. 

혹시 어떤 문제인지 아시는 분 계신가요 ??

## Resource
- https://jusung.gitbook.io/the-swift-language-guide/language-guide/25-access-control

**Comment by YoonAh-dev** [comment]
enum이 private으로 설정되어 있다면 해당 enum를 사용하는 객체도 private으로 범위가 설정되어야 합니다.

접근제한자는 open > public > internal > fileprivate > private 순으로 구성이 되어 있습니다.
만약 public으로 enum를 설정해둔다면 enum 객체는 public부터 private까지 모두 사용이 가능합니다. 하지만 반대로 private으로 설정해뒀다면 private 이외에는 객체의 접근제한자를 설정할 수 없습니다. enum이 private인데 해당 타입을 받은 객체가 public으로 설정되어 있는게 말이 안되니깐요! 

같은 의미로 internal로 설정되어 있으면 public으로 객체를 사용할 수 없습니다.
<img width="792" alt="스크린샷 2022-08-10 오후 2 18 30" src="https://user-images.githubusercontent.com/55099365/183821565-e9701e8a-7ebe-4062-be13-fcec1677d419.png">

enum을 private으로 설정했다는건 해당 enum를 현재 class에서만 사용한다는 뜻인데, class밖에 있는 다른 class, struct들이 해당 변수에 접근할 수 있다는 건 애초에 해당 enum를 다른 class, struct도 봐야한다는 뜻이기 때문에 enum를 internal로 설정해주는 게 맞겠죠? 또한 private으로 설정된 변수를 다른 클래스에서 접근한다는 건 캡슐화, 은닉화를 위배한 것으로 볼 수 있습니다.
<img width="895" alt="스크린샷 2022-08-10 오후 2 25 57" src="https://user-images.githubusercontent.com/55099365/183822353-e1cc8f37-9978-4e8b-be2f-ff92e885de86.png">

**Reply by MMMIIIN** [reply]
답변 감사합니다 듀나 ! ^______^ 😆

제가 의문을 가졌던 부분은 변수를 `private var` 로 선언해두고 외부에서 그 값을 변경하기 위해 함수를 하나 생성해서 컨트롤하는데 `enum`은 그게 되지않는게 의문이었습니다 !

예를들어 
```swift
class CalendarView: UIView {
    private var selectStartDate = Date()
    ...
}
```

위와 같이 한 `class` 내부에 `private` 으로 변수를 생성하면 외부에서 접근을 못하잖아요 ??
```swift
class CalendarView: UIView {
    private var selectStartDate = Date()

    ...
    ...

    func changeSelectDate() {
        self.selectStartDate = Date() // 예시
    }
}
```
그래서 위와 같이 해당 변수에 접근할 수 있는 함수를 `internal`로 선언하죠

이때는 위의 `enum` 문제와 같은 현상이 발생하지 않아서, `enum`을 사용할 때도 비슷한 맥락이겠거니 생각했는데 아니었단말이죠..??

이건 어떤 차이일까요 ?? `enum`의 특성일까요...??

**Reply by YoonAh-dev** [reply]
changeSelectDate의 경우에는 내부에 있는 selectStartDate를 다른 class가 몰라도 괜찮은 상황입니다. 그냥 changeSelectDate에 매개변수로 Date 타입의 값을 넣으면 현재 class 내부에서 알아서 selectStartDate에 Date값을 넣어주니깐요!

만약 `func changeSelectEditMode(_ mode: EditMode)`를 만들어서 사용한다고 하면 결국 mode에 있는 EditMode값을 다른 class가 알아야지만 해당 값을 사용할 수 있어서 EditMode를 internal로 설정해줘야 할겁니다. enum 타입은 `selectStartDate`에 매개변수로 들어가는 `Date`타입과는 다르게 사용자가 만든 타입이기 때문에 다른 class들이 알 수 없게 private으로 설정해두면 당연히 해당 타입을 모르기 때문에 다른 class들은 사용이 불가능하겠죠? 왜냐면 Date타입은 Foundation에서 제공해주는 타입이기 때문에 모든 class가 다 알고 있지만 EditMode는 CalendarView class만 알고 있는 타입이기 때문이죠.

enum의 특성이라기 보다는 enum 타입을 class에 숨겨두었기 때문에 생기는 일인거 같습니다. 만약 모든 class들이 해당 enum타입을 알고 싶게 만들고 싶다면 internal로 설정해주던지, 전역 변수로 설정해두는게 맞는 방식일 거 같습니다.

**Reply by MMMIIIN** [reply]
아하 그렇군요 ! 
핵심이 enum은 내가 만든 타입이다 군요 !
뭔가 집안에 숨겨진 물건을 찾기위해 `internal` 적인 사람(함수)를 만들어서 접근한다고 생각했는데, 일반적인 상황은 모두가 어떤 물건(타입)인지 알고 있었던 거군요 !
똑같은 방식으로 접근은 가능할 예정이지만 `private`로 정의된 타입을 모르기 때문에 에러가 나는군요 !
현재 class에서만 생각해서 뭐가 문젠지 몰랐네요..
설명이 좀 이상하지만 이해했습니다 ! 감사합니다 ㅎㅎ ^________^

듀나의 설명을 들으니 굉장히 멍청한 질문이었던 것 같군요.. ㅎㅎ (머쓱) 🙃

듀나멋져 ^______^

**Reply by YoonAh-dev** [reply]
케미멋져! ^_____^ 화이팅입니다
